hal/src/nRF52840/core_hal.c

```diff
@@ -464,11 +486,229 @@ void HAL_Core_Enter_Bootloader(bool persist) {
+        return SYSTEM_ERROR_NOT_ALLOWED;
+    }
+
+    for (unsigned i = 0; i < pins_count; i++) {
+        if (pins[i] >= TOTAL_PINS) {
+            return SYSTEM_ERROR_NOT_ALLOWED;
+        }
+    }
+
+    for (int i = 0; i < mode_count; i++) {
+        switch(mode[i]) {
+            case RISING:
+            case FALLING:
+            case CHANGE:
+                break;
+            default:
+                return SYSTEM_ERROR_NOT_ALLOWED;
+        }
+    }
+
+    SysTick->CTRL &= ~SysTick_CTRL_ENABLE_Msk;
+
+    // Detach USB
+    HAL_USB_Detach();
+
+    // Disable RTC2
+    nrf5AlarmDeinit();
+
+    // Disable external flash
+    hal_exflash_uninit();
+
+    // Flush all USARTs
+    for (int usart = 0; usart < TOTAL_USARTS; usart++) {
+        if (HAL_USART_Is_Enabled(usart)) {
+            HAL_USART_Flush_Data(usart);
+        }
+    }
+
+    // Disable all interrupt, __disable_irq()
+    uint8_t dummy = 0;
+    uint32_t err_code = sd_nvic_critical_region_enter(&dummy);
+    SPARK_ASSERT(err_code == NRF_SUCCESS);
+
+    wakeup_info.source = STOP_MODE_EXIT_CONDITION_NONE;
+
+    bool hfclk_resume = false;
+    if (nrf_drv_clock_hfclk_is_running()) {
+        hfclk_resume = true;
+        nrf_drv_clock_hfclk_release();
+        while (nrf_drv_clock_hfclk_is_running()) {
+            ;
+        }
+    }
+
+    uint32_t exit_conditions = 0x00;
+
+    // Suspend all GPIOTE interrupts
+    HAL_Interrupts_Suspend();
+
+    for (int i = 0; i < pins_count; i++) {
+        pin_t wakeUpPin = pins[i];
+        InterruptMode edgeTriggerMode = (i < mode_count) ? mode[i] : mode[mode_count - 1];
+
+        PinMode wakeUpPinMode = INPUT;
+        // Set required pinMode based on edgeTriggerMode
+        switch(edgeTriggerMode) {
+            case RISING: {
+                wakeUpPinMode = INPUT_PULLDOWN;
+                break;
+            }
+            case FALLING: {
+                wakeUpPinMode = INPUT_PULLUP;
+                break;
+            }
+            case CHANGE:
+            default:
+                wakeUpPinMode = INPUT;
+                break;
+        }
+
+        HAL_Pin_Mode(wakeUpPin, wakeUpPinMode);
+        HAL_InterruptExtraConfiguration irqConf = {0};
+        irqConf.version = HAL_INTERRUPT_EXTRA_CONFIGURATION_VERSION_2;
+        irqConf.IRQChannelPreemptionPriority = 0;
+        irqConf.IRQChannelSubPriority = 0;
+        irqConf.keepHandler = 1;
+        irqConf.keepPriority = 1;
+        HAL_Interrupts_Attach(wakeUpPin, wakeup_gpiote_handler, (void*)(uint32_t)i, edgeTriggerMode, &irqConf);
+
+        exit_conditions |= STOP_MODE_EXIT_CONDITION_PIN;
+    }
+
+    // Configure RTC wake-up
+    if (seconds > 0) {
+        wakeup_from_rtc(seconds);
+
+        exit_conditions |= STOP_MODE_EXIT_CONDITION_RTC;
+    }
+
+    // Enable all interrupt, __enable_irq()
+    err_code = sd_nvic_critical_region_exit(dummy);
```

Comment:
Should not be enabled here. They should be disabled almost until exiting this function after the wakeup.